redshift/redshift.go

```diff
@@ -479,6 +482,48 @@ func (r *Redshift) Copy(tx *sql.Tx, f s3filepath.S3File, delimiter string, creds
 	return err
 }
 
+// UpdateLatencyInfo updates the latency table with the current time to indicate
+// that the table data has been updated
+func (r *Redshift) UpdateLatencyInfo(tx *sql.Tx, table Table) error {
+	dest := fmt.Sprintf("%s.%s", table.Meta.Schema, table.Name)
+
+	// Insert a row for the latencies table if it doesn't already exist.
+	// We do this outside of the transaction, so that we can get a row-level lock with the next command.
+	_, err := r.ExecContext(r.ctx, fmt.Sprintf(
+		`INSERT INTO latencies (name) (
+				SELECT name FROM latencies
+			UNION
+				SELECT '%s' AS name
+			EXCEPT
+				SELECT name FROM latencies
+		)`, dest))
+
+	// Get the last latency value out of the table, for logging. This doesn't have to be in the transaction.
+	latencyQuery := fmt.Sprintf("SELECT last_update FROM latencies WHERE name = '%s'", dest)
+	var t pq.NullTime
+	err = r.QueryRowContext(r.ctx, latencyQuery).Scan(&t)
+	// this will either return a value or null if no data, rather than no rows, because we inserted earleir
+	if err != nil {
+		return fmt.Errorf("error scanning latency table for %s: %s", dest, err)
+	}
+
+	// Update the latency table with the current timestamp, for the last run.
+	// We *do* want this one inside the transaction!
+	_, err = tx.ExecContext(r.ctx, fmt.Sprintf(
+		"UPDATE latencies SET last_update = current_timestamp WHERE name = '%s'",
+		dest))
+	if err != nil {
+		return err
+	}
+
+	logger.GetLogger().InfoD("analytics-run-latency", kvlogger.M{
+		"destination":  dest,
+		"previous_run": t.Time,
```

Comment:
Why do we want the previous time? Won't that be in the event stream anyway?

Reply:
Oh, I guess this gives us the actual current_timestamp rather than the log time.

Reply:
We could select it out of the transaction too if we want?

Reply:
Oh, but we don't know if this ends up succeeding until later, haha